### PR TITLE
GET /planters api has duplicated records #186

### PR DESCRIPTION
**Pull request by dayaneoli**
@dadiorchen @jeremydthomas. The intention is to use the set() object to avoid repeated ID problem (GET /planters api has duplicated records #186).

plantersRouter.ts

rt express from 'express';
import Joi from 'joi';
import FilterOptions from 'interfaces/FilterOptions';
import { handlerWrapper } from './utils';
import PlanterRepository from '../infra/database/PlanterRepository';
import Session from '../infra/database/Session';
import PlanterModel from '../models/Planter';

const router = express.Router();
type Filter = Partial<{ planter_id: number; organization_id: number }>;

router.get(
  '/featured',
  handlerWrapper(async (req, res) => {
    const repo = new PlanterRepository(new Session());
    const result = await PlanterModel.getFeaturedPlanters(repo)({ limit: 10 });
    res.send({
      planters: result.map((planter) => ({
        ...planter,
        links: PlanterModel.getPlanterLinks(planter),
      })),
    });
    res.end();
  }),
);

router.get(
  '/:id',
  handlerWrapper(async (req, res) => {
    Joi.assert(req.params.id, Joi.number().required());
    const repo = new PlanterRepository(new Session());
    const exe = PlanterModel.getById(repo);
    const result = await exe(req.params.id);
    result.links = PlanterModel.getPlanterLinks(result);
    _**/*  const mySet = {repo}  The intention is to use the set() object to avoid repeated ID
    new Set(mySet); */**_
    res.send(result);
    res.end(); 
  }),
);

router.get(
  '/',
  handlerWrapper(async (req, res) => {
    Joi.assert(
      req.query,
      Joi.object().keys({
        keyword: Joi.string(),
        organization_id: Joi.number().integer().min(0),
        limit: Joi.number().integer().min(1).max(1000),
        offset: Joi.number().integer().min(0),
        order_by: Joi.string(),
        order: Joi.string().valid('asc', 'desc', 'ASC', 'DESC'),
      }),
    );
    const {
      limit = 20,
      offset = 0,
      organization_id,
      keyword,
      order_by = null,
      order = 'asc',
    } = req.query;
    const options: FilterOptions = { limit, offset };
    if (order_by) {
      options.orderBy = { column: order_by, direction: order };
    }
    const repo = new PlanterRepository(new Session());
    const filter: Filter = {};
    if (organization_id) {
      filter.organization_id = organization_id;
    }

    if (keyword !== undefined) {
      const result = await PlanterModel.getByName(repo)(keyword, options);
      res.send({
        total: await PlanterModel.countByName(repo)(keyword),
        offset,
        limit,
        planters: result.map((planter) => ({
          ...planter,
          links: PlanterModel.getPlanterLinks(planter),
        })),
      });
      res.end();
    } else {
      const result = await PlanterModel.getByFilter(repo)(filter, options);
      res.send({
        total: await PlanterModel.countByFilter(repo)(filter),
        offset,
        limit,
        planters: result.map((planter) => ({
          ...planter,
          links: PlanterModel.getPlanterLinks(planter),
        })),
      });
      res.end();
    }
  }),
);

export default router;

### CODE REVIEW

**Comment by dayaneoli** [comment]
@dadiorchen @jeremydthomas. The intention is to use the set() object to avoid repeated ID problem (GET /planters api has duplicated records https://github.com/Greenstand/treetracker-query-api/issues/186).

File: .github/workflows/treetracker-query-api-build-deploy-dev.yml [hunk 1]
```diff
@@ -20,10 +20,10 @@ jobs:
       - uses: actions/checkout@v2
         with:
           ref: ${{ github.event.inputs.git-tag }}
-      - name: Use Node.js 16.x
+      - name: Use Node.js 18.x #semantic-release
         uses: actions/setup-node@v1
         with:
-          node-version: '16.x'
+          node-version: '18.x'
```
Comment (dadiorchen) [inline .github/workflows/treetracker-query-api-build-deploy-dev.yml:26]: If we need to upgrade the node version, I suggest we put it into a dedicated PR to do it, the smaller and more focused PR, the better

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "treetracker-query-api",
-  "version": "1.59.0",
+  "version": "1.62.1",
```
Comment (dadiorchen) [inline package.json]: This version number should be automatically updated by our semantic-robot, so it shouldn't be modified in a PR, is there any difficulty for you to avoid this?

File: server/app.ts [hunk 3]
```diff
@@ -59,16 +60,21 @@ app.use(express.json()); // parse application/json
...
+app.use('/v2/wallets', walletsRouter);
 app.use('/transactions', transactionsRouter);
 app.use('/tokens', tokensRouter);
+app.use('/v2/tokens', tokensRouter);
 app.use('/v2/captures', capturesRouter);
 app.use('/raw-captures', rawCapturesRouter);
-app.use('/grower-accounts', growerAccountsRouter);
+app.use('/v2/growers', growerAccountsRouter);
 app.use('/v2/trees', treesRouterV2);
 app.use('/bounds', boundsRouter);
```
Comment (dadiorchen) [inline server/app.ts:79]: @jeremydthomas I'm confused by this, is this your code for the v2? Why it shows up here?

**Review by dadiorchen — CHANGES_REQUESTED** [review]

File: server/app.ts (hunk 3 above)
Comment (jeremydthomas) [reply server/app.ts:79]: @dadiorchen that is a good question, I have no idea. I didnt touch her code at all. It shouldnt be there.


EDIT: it looks like she started on this long before those changes were implemented. In fact I know she did, she has been working on this quite awhile. Im guessing these were changes were from the new code getting into merged into the old.

Comment (dadiorchen) [reply server/app.ts:79]: @dayaneoli could you do some clean up on this PR? IMO, we can checkout the main branch and `git cherry-pick` your commit (valid part to solve the problem) then raise a new PR?

File: .github/workflows/treetracker-query-api-build-deploy-dev.yml (hunk 1 above)
Comment (dayaneoli) [reply .github/workflows/treetracker-query-api-build-deploy-dev.yml:26]: The PR wasn't for that. My virtual machine that must have upgraded to work.

File: package.json (hunk 2 above)
Comment (dayaneoli) [reply package.json]: My virtual machine that must have updated. If you can tell me please how do I do, not to climb the PR.

File: server/app.ts (hunk 3 above)
Comment (dayaneoli) [reply server/app.ts:79]: I forkei the repository and put the object my set(). In the rest of the code I didn't change anything.

Comment (dayaneoli) [reply server/app.ts:79]: I'll try to do that and then a new PR. Thanks.
